runtime/src/iree/hal/drivers/hip/pending_queue_actions.c

```diff
@@ -1064,11 +1312,147 @@ static void iree_hal_hip_worker_wait_pending_work_items(
+    return -1;
+  }
+
+  while (true) {
+    iree_notification_await(
+        &completion_area->state_notification,
+        (iree_condition_fn_t)
+            iree_hal_hip_completion_has_incoming_request_or_error,
+        completion_area, iree_infinite_timeout());
+
+    IREE_TRACE_ZONE_BEGIN(z0);
+    // Immediately flip the state to idle waiting if and only if the previous
+    // state is workload pending. We do it before processing ready list to make
+    // sure that we don't accidentally ignore new workload pushed after done
+    // ready list processing but before overwriting the state from this worker
+    // thread. Also we don't want to overwrite other exit states. So we need to
+    // perform atomic compare and exchange here.
+    iree_hal_hip_worker_state_t prev_state =
+        IREE_HAL_HIP_WORKER_STATE_WORKLOAD_PENDING;
+    iree_atomic_compare_exchange_strong_int32(
+        &completion_area->worker_state, /*expected=*/&prev_state,
+        /*desired=*/IREE_HAL_HIP_WORKER_STATE_IDLE_WAITING,
+        /*order_succ=*/iree_memory_order_acq_rel,
+        /*order_fail=*/iree_memory_order_acquire);
+
+    int32_t worker_state = iree_atomic_load_int32(
+        &completion_area->worker_state, iree_memory_order_acquire);
+    // Exit if HIP callbacks have posted any errors.
+    if (IREE_UNLIKELY(worker_state == IREE_HAL_HIP_WORKER_STATE_EXIT_ERROR)) {
+      iree_hal_hip_completion_wait_pending_completion_items(completion_area);
+      IREE_TRACE_ZONE_END(z0);
+      return -1;
+    }
+    // Check if we received request to stop processing and exit this thread.
+    bool should_exit =
+        (worker_state == IREE_HAL_HIP_WORKER_STATE_EXIT_REQUESTED);
+
+    iree_status_t status = iree_ok_status();
+    while (true) {
+      iree_hal_hip_atomic_slist_completion_t* entry =
+          iree_hal_hip_completion_slist_pop(worklist);
+      if (!entry) break;
+
+      const iree_hal_hip_dynamic_symbols_t* symbols = completion_area->symbols;
+      IREE_TRACE_ZONE_BEGIN_NAMED(z1, "hipEventSynchronize");
+      hipError_t result = symbols->hipEventSynchronize(entry->event);
+      IREE_TRACE_ZONE_END(z1);
+      if (IREE_UNLIKELY(result != hipSuccess)) {
+        // Let common destruction path take care of destroying the worklist.
+        // When we know all host stream callbacks are done and not touching
+        // anything.
+        iree_hal_hip_completion_slist_push(worklist, entry);
+        break;
+      }
+      entry->callback(entry->user_data);
+
+      if (IREE_UNLIKELY(entry->created_event)) {
+        IREE_HIP_IGNORE_ERROR(symbols, hipEventDestroy(entry->event));
+      }
+      iree_allocator_free(completion_area->host_allocator, entry);
+
+      // Now we fully executed and cleaned up this entry. Decrease the work
+      // items counter.
+      iree_hal_hip_queue_decrement_completion_count(completion_area);
+    }
+
+    if (IREE_UNLIKELY(!iree_status_is_ok(status))) {
```

Comment:
`status` is not actually wired up here? we should be able to update `status` at line 1398 for example.